### PR TITLE
販売データインポートモジュール改修

### PR DESCRIPTION
**Pull request by TimLai125**

### CODE REVIEW

**Comment by yostashiro** [comment]
@TimLai125 

I get this error when I try creating a record for sale.import.default.

image

I think this field should be removed:
https://github.com/rfhk/sts-custom/blob/e6f190d7cce5d899feda54a7d9af5999558fa4f4/sale_order_import/models/sale_import_default.py#L22-L28

But what is your assumption on different settings of `invoice_policy` (`order` or `delivery`) in product?  Is this covered in your test case?  Note that there used to be `order_policy` field in sale.order up to V8 which was defaulted from sales configuration, however the concept was replaced by `invoice_policy` in product.

File: sale_order_import/wizard/import_sale.py [hunk 1]
```diff
@@ -223,167 +520,190 @@ def import_sale_data(self):
...
+
+            #  column validation
+            missing_columns = list(set(FIELDS_TO_IMPORT) - set(sheet_fields))
+            if missing_columns:
+                raise Warning(
+                    _('Following columns are missing: \n %s'
+                      % ('\n'.join(missing_columns))
+                      )
+                )
+
```
Comment (yostashiro) [inline sale_order_import/wizard/import_sale.py]: Below lines are too redundant.  Replace them by:
```
        order_group = sheet_fields.index('Group')
        partner_name = sheet_fields.index('Customer')
        partner_tel = sheet_fields.index('Customer Phone/Mobile')
        product_id = sheet_fields.index('Line Product')
        line_name = sheet_fields.index('Line Description')
        price_unit = sheet_fields.index('Line Unit Price')
        product_qty = sheet_fields.index('Line Qty')
        taxes_id = sheet_fields.index('Line Tax')
        notes = sheet_fields.index('Notes')
        pricelist_id = sheet_fields.index('Pricelist')
        warehouse_id = sheet_fields.index('Warehouse')
        team_id = sheet_fields.index('Team')
        carrier_id = sheet_fields.index('Carrier')
```

Or there may be a smarter way to do this.

File: sale_order_import/wizard/import_sale.py [hunk 2]
```diff
@@ -2,202 +2,494 @@
...
+                          'price_unit': so_line['price_unit'],
+                          'state': so_line['state'],
+                          'tax_id': [(6, 0, so_line['tax_id'])],
+                          'order_id': order_id.id}
         return self.env['sale.order.line'].create(orderline_vals)
 
-
     @api.multi
     def import_sale_data(self):
+        """Import sale data"""
```
Comment (yostashiro) [inline sale_order_import/wizard/import_sale.py]: Remove this comment as it doesn't add any more info.

File: sale_order_import/wizard/import_sale.py [hunk 3]
```diff
@@ -2,202 +2,494 @@
...
+        ctx = self._context.copy()
         for line in self:
```
Comment (yostashiro) [inline sale_order_import/wizard/import_sale.py]: Remove this loop and use `self.ensure_one()` instead.

**Comment by TimLai125** [comment]
@yostashiro 

> But what is your assumption on different settings of invoice_policy (order or delivery) in product? Is this covered in your test case? Note that there used to be order_policy field in sale.order up to V8 which was defaulted from sales configuration, however the concept was replaced by invoice_policy in product.

My assumption is all the products are configured with "order" (Ordered quantities) as the invoice_policy (Invoicing Policy). Hence I did not add to the test cases

**Comment by yostashiro** [comment]
> My assumption is all the products are configured with "order" (Ordered quantities) as the invoice_policy (Invoicing Policy). Hence I did not add to the test cases

How valid is this assumption?  What if there are exception cases?  No chances of confusion?

**Comment by TimLai125** [comment]
> How valid is this assumption? What if there are exception cases? No chances of confusion?

Ok, I will handle the other invoice_policy situation as well.

**Pull request merged**